src/main/java/org/folio/circulation/rules/cache/CirculationRulesCache.java

```diff
@@ -1,208 +1,124 @@
+    if (log.isInfoEnabled()) {
+      log.info("circulationRules = {}", circulationRules.encodePrettily());
+    }
+
+    return circulationRules.getString("rulesAsText");
+  }
+
+  public CompletableFuture<Result<Rules>> reloadRules(String tenantId, String rulesAsText) {
+    log.info("Reloading rules for tenant {}", tenantId);
 
         if (isBlank(rulesAsText)) {
           log.info("Rules text is blank for tenant {}", tenantId);
           return completedFuture(failed(new ServerErrorFailure(
             "Cannot apply blank circulation rules")));
         }
 
-        if (rules.rulesAsText.equals(rulesAsText) && !rebuildNeeded(tenantId, rules)) {
-          log.info("Rules have not changed for tenant {} and rebuild is not needed",
-            tenantId);
-          return ofAsync(() -> rules);
-        }
+        String droolsText = Text2Drools.convert(rulesAsText);
+        Drools drools =  new Drools(tenantId, droolsText);
+        log.info("rulesAsDrools = {}", droolsText);
 
-        rules.rulesAsText = rulesAsText;
-
-        rules.rulesAsDrools = Text2Drools.convert(rulesAsText);
-        log.info("rulesAsDrools = {}", rules.rulesAsDrools);
-
-        rules.drools = new Drools(tenantId, rules.rulesAsDrools);
-        rules.rebuildTimestamp = System.currentTimeMillis();
+        Rules rules = new Rules(rulesAsText, droolsText, drools, System.currentTimeMillis());
         log.info("Done building Drools object for tenant {}", tenantId);
 
+        rulesMap.put(tenantId, rules);
+
         return ofAsync(() -> rules);
-      }));
   }
 
   public CompletableFuture<Result<ExecutableRules>> getExecutableRules(String tenantId,
     CollectionResourceClient circulationRulesClient) {
 
     return getDrools(tenantId, circulationRulesClient)
       .thenApply(r -> r.map(drools ->
-        new ExecutableRules(rulesMap.get(tenantId).rulesAsText, drools)));
+        new ExecutableRules(rulesMap.get(tenantId).getRulesAsText(), drools)));
   }
 
   public CompletableFuture<Result<Drools>> getDrools(String tenantId,
     CollectionResourceClient circulationRulesClient) {
 
-    log.info("Getting Drools for tenant {}", tenantId);
+    log.debug("getDrools:: parameters: tenantId, circulationRulesClient: Getting Drools for tenant {}", tenantId);
 
     final CompletableFuture<Result<Drools>> cfDrools = new CompletableFuture<>();
-    Rules rules = rulesMap.get(tenantId);
-
-    if (isCurrent(tenantId, rules)) {
-      log.info("Rules for tenant {} are current, returning immediately: {}", tenantId,
-        rules.rulesAsText);
-
-      cfDrools.complete(succeeded(rules.drools));
-
-      if (reloadNeeded(tenantId, rules)) {
-        log.info("Need to reload rules for tenant {}", tenantId);
-
-        rules.reloadInitiated = true;
-        reloadRules(tenantId, rules, circulationRulesClient)
-          .thenCompose(r -> r.after(updatedRules -> ofAsync(() -> updatedRules.drools)));
-      }
 
+    if (rulesExist(tenantId)) {
```

Comment:
```suggestion
    if (tenantId != null && rulesExist(tenantId) {
```
I think it makes sense to include a null check for tenantId since ConcurrentHashMap throws a NullPointerException if a null key is provided